### PR TITLE
`Development`: Convert classes into records

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The main motivation for this change is to take advantage of records, which were introduced in Java 14 and formally standardised in Java 16. Records provide a more concise and readable way to define immutable data carriers, which is the main purpose of DTOs.
- **Conciseness**: Records significantly reduce boilerplate code. DTOs often require multiple lines of code to define fields, constructors, accessors, equals(), hashCode(), and toString() methods. With records, all of these are automatically provided by the compiler.
- **Immutability**: Records are immutable by default. This aligns with the design principles of DTOs, which should represent data snapshots without exposing mutability.
- **Readability**: By removing boilerplate code, records make the DTO definitions cleaner and more focused on the actual data they carry, improving readability and maintainability.
- **Performance**: The auto-generated methods in records are optimized by the compiler, potentially offering better performance characteristics for equals(), hashCode(), and toString().

### Description

- Refactored multiple DTO classes in the project to use Java records.
- Added an architecture test that prevents the addition of new DTO classes.
- Updated relevant test cases to reflect the changes.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Tutor
- 1 Empty Course

- This PR is best tested locally, as you have access to the server's logs there.
- Look out for Bad Requests and Nullpointer Exceptions. They indicate that a Collection type that was never null by default before might be able to be null now.
- Given the changed DTOs, try to break the communication between the client and the server (e.g. Write a complaint as a student and then answer it as an instructor to test the AssessmentUpdateDTO etc.). Please make sure to add what you've tested into your review description.
- Likely edge cases include empty collections (e.g. an AssessmentUpdate with no feedbacks). The classes may have set these values to an empty collection by default, but the records do not, leading to potential errors when iterating directly over collections elsewhere in the code without first checking them for nullity.

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenResponseDTO.java [hunk 1]
```diff
@@ -1,25 +1,14 @@
...
-
-    public GitLabPersonalAccessTokenResponseDTO() {
-        // default constructor for Jackson
-    }
-
-    public String getToken() {
-        return token;
-    }
+public record GitLabPersonalAccessTokenResponseDTO(@JsonProperty String name, @JsonProperty("user_id") Long userId, @JsonProperty("scopes") String[] scopes,
+        @JsonProperty("expires_at") Date expiresAt, @JsonProperty String token) {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenResponseDTO.java:12]:
```suggestion
public record GitLabPersonalAccessTokenResponseDTO(String name, @JsonProperty("user_id") Long userId, String[] scopes,
        @JsonProperty("expires_at") Date expiresAt, String token) {
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenRequestDTO.java [hunk 2]
```diff
@@ -8,59 +8,7 @@
 
 @JsonIgnoreProperties(ignoreUnknown = true)
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
-public class GitLabPersonalAccessTokenRequestDTO {
+public record GitLabPersonalAccessTokenRequestDTO(@JsonProperty String name, @JsonProperty("user_id") Long userId, @JsonProperty("scopes") String[] scopes,
+        @JsonProperty("expires_at") Date expiresAt) {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenRequestDTO.java:12]: I think we can just remove the `@JsonProperty` annotations since they don't do anything, right?
```suggestion
public record GitLabPersonalAccessTokenRequestDTO(String name, @JsonProperty("user_id") Long userId, String[] scopes,
        @JsonProperty("expires_at") Date expiresAt) {
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabCommitDTO.java [hunk 3]
```diff
@@ -10,111 +10,17 @@
...
-        return removed;
-    }
-
-    public void setRemoved(List<String> removed) {
-        this.removed = removed;
+    // make sure the lists are never null
+    public GitLabCommitDTO {
+        added = (added != null) ? added : new ArrayList<>();
+        modified = (modified != null) ? modified : new ArrayList<>();
+        removed = (removed != null) ? removed : new ArrayList<>();
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabCommitDTO.java:20]: We should probably adjust the code utilizing this DTO to also accept empty Arrays (if it makes sense), since all other DTOs also send null values for empty collections.

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java [hunk 4]
```diff
@@ -1119,15 +1112,18 @@ public ExamChecklistDTO getStatsForChecklist(Exam exam, boolean isInstructor) {
...
-        examChecklistDTO.setNumberOfTotalParticipationsForAssessment(totalNumberOfParticipationsForAssessment);
+
         boolean existsUnassessedQuizzes = submissionRepository.existsUnassessedQuizzesByExamId(exam.getId());
-        examChecklistDTO.setExistsUnassessedQuizzes(existsUnassessedQuizzes);
         boolean existsUnsubmittedExercises = submissionRepository.existsUnsubmittedExercisesByExamId(exam.getId());
-        examChecklistDTO.setExistsUnsubmittedExercises(existsUnsubmittedExercises);
-        return examChecklistDTO;
+
+        // For non-instructors, consider what limited information they should receive and adjust accordingly
+        return new ExamChecklistDTO(null, null, totalNumberOfAssessmentsFinished, totalNumberOfParticipationsForAssessment, null, null, null, null, null, existsUnassessedQuizzes,
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java:1124]: would it make sense to add another constructor/method to the DTO that only takes the data instructors should see as input?

File: src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java [hunk 5]
```diff
@@ -443,7 +443,7 @@ private Conversation mayUpdateOrDeleteMessageElseThrow(Post existingMessagePost,
      * @return list of similar posts
      */
     public List<Post> getSimilarPosts(Long courseId, Post post) {
-        PostContextFilter postContextFilter = new PostContextFilter(courseId);
+        PostContextFilter postContextFilter = new PostContextFilter(courseId, null, null, null, null, false, false, false, null, null);
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java:446]: see above comment, another constructor/method could make the code more readable.

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/AssessmentUpdateDTO.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * @param complaintResponse the corresponding complaint response
+ * @param assessmentNote    the assessment note
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record AssessmentUpdateDTO(List<Feedback> feedbacks, ComplaintResponse complaintResponse, @Nullable String assessmentNote) implements AssessmentUpdate {
+
+    /*
+     * Checks if the feedbacks list is null. If it is, it assigns it a default value of an empty list.
+     */
+    public AssessmentUpdateDTO {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/AssessmentUpdateDTO.java:24]: Again, we should check if we can just adjust the surrounding code to support null values.

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/PostContextFilter.java [hunk 7]
```diff
@@ -8,115 +8,22 @@
 import de.tum.in.www1.artemis.domain.metis.PostSortCriterion;
 
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
-public class PostContextFilter {
+public record PostContextFilter(@NotBlank Long courseId, long[] courseWideChannelIds, Long plagiarismCaseId, Long conversationId, String searchText, Boolean filterToUnresolved,
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/PostContextFilter.java:11]: Should we rename this class to `PostContextFilterDTO`? (And also its occurrences in variables to "...DTO")?

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamLiveEventDTO.java [hunk 8]
```diff
@@ -20,52 +17,6 @@
...
-        if (this == o) {
-            return true;
-        }
-        if (o == null || getClass() != o.getClass()) {
-            return false;
-        }
-        ExamLiveEventDTO that = (ExamLiveEventDTO) o;
-        return Objects.equals(id, that.id) && Objects.equals(createdBy, that.createdBy) && Objects.equals(createdDate, that.createdDate);
-    }
+public interface ExamLiveEventDTO {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamLiveEventDTO.java:20]: since this does nothing, should we remove it?

File: src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java [hunk 9]
```diff
@@ -29,9 +29,6 @@ public void setTextContent(String textContent) {
 
     @Override
     public ExamAttendanceCheckEventDTO asDTO() {
-        var dto = new ExamAttendanceCheckEventDTO();
-        super.populateDTO(dto);
-        dto.setText(textContent);
-        return dto;
+        return new ExamAttendanceCheckEventDTO(this.getId(), this.getCreatedBy(), this.getCreatedDate(), textContent);
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java:32]: IMO it would make more sense to add an `public static of(ExamAttendenceCheck e)` method to the `ExamAttendenceCheckDTO`. We have these methods in multiple DTOs already and it would make the code more consistent.

File: src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java [hunk 10]
```diff
@@ -202,6 +202,18 @@ void testJSONImplementations() {
                 .check(allClasses);
     }
 
+    @Test
+    void testDTOImplementations() {
+        var dtoRecordRule = classes().that().haveSimpleNameEndingWith("DTO").should().beRecords().andShould().beAnnotatedWith(JsonInclude.class)
+                // .andShould(useJsonIncludeNonEmpty())
+                .because("All DTOs should be records and annotated with @JsonInclude(JsonInclude.Include.NON_EMPTY)");
```
Comment (rstief) [inline src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java:209]: Why was this line commented out? If we want to re-add these checks we should at least add a TODO. if it should stay removed we should adjust the failure message accordingly.
```suggestion
        var dtoRecordRule = classes().that().haveSimpleNameEndingWith("DTO").should().beRecords().andShould().beAnnotatedWith(JsonInclude.class)
                // .andShould(useJsonIncludeNonEmpty())
                .because("All DTOs should be records and annotated with @JsonInclude");
```

**Review by rstief — DISMISSED** [review]
Code LGTM overall. I have some suggestions for improvements, but those can also be added in a follow-up.

Key Points (for the future):
- IMO DTOs should support null values for collections, so we should remove the compact constructors setting empty collections
- IMO we should use `of(...)` methods in the DTO rather than `asDTO(...)` methods in the entity

File: src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java (hunk 9 above)
Comment (JohannesWt) [reply src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java:32]: Additionally to the consistency point of Raphael, it makes sense to move this convertion method into the DTO as several DTOs for one class may exists. Therefore the concrete mapping should be handled by the individual DTO to separate concerns.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenListResponseDTO.java [hunk 11]
```diff
@@ -6,30 +6,10 @@
...
-    }
-
-    public Long getId() {
-        return id;
-    }
-
-    public void setId(Long id) {
-        this.id = id;
-    }
+public record GitLabPersonalAccessTokenListResponseDTO(@JsonProperty Long id) {
```
Comment (JohannesWt) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenListResponseDTO.java:14]: Why do we need `@JsonProperty` here if we don't pass a different json-key value?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenRequestDTO.java [hunk 12]
```diff
@@ -8,59 +8,7 @@
 
 @JsonIgnoreProperties(ignoreUnknown = true)
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
-public class GitLabPersonalAccessTokenRequestDTO {
+public record GitLabPersonalAccessTokenRequestDTO(@JsonProperty String name, @JsonProperty("user_id") Long userId, @JsonProperty("scopes") String[] scopes,
```
Comment (JohannesWt) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenRequestDTO.java:11]: Same here - why do we need `@JsonProperty` for name?

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java (hunk 4 above)
Comment (JohannesWt) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java:1124]: Then maybe it makes sense to mark those values as `@Nullable` in the DTO? (use the jakarta Nullable)

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamLiveEventDTO.java (hunk 8 above)
Comment (JohannesWt) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamLiveEventDTO.java:20]: but why are there no fields like in the old class DTO? (id, createdBy, createdDate)?

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamWideAnnouncementEventDTO.java [hunk 13]
```diff
@@ -1,39 +1,13 @@
...
-            return false;
-        }
-        if (!super.equals(o)) {
-            return false;
-        }
-        ExamWideAnnouncementEventDTO that = (ExamWideAnnouncementEventDTO) o;
-        return Objects.equals(text, that.text);
-    }
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record ExamWideAnnouncementEventDTO(Long id, String createdBy, Instant createdDate, String text) implements ExamLiveEventDTO {
```
Comment (JohannesWt) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamWideAnnouncementEventDTO.java:11]: id, createdBy and createdDate should be handled by interface

**Review by JohannesWt — APPROVED** [review]
Code looks good - few (maybe optional/debatable) changes explained in the comments

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamLiveEventDTO.java (hunk 8 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/examevent/ExamLiveEventDTO.java:20]: It is used in some cases as response type which makes sense to me,

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenRequestDTO.java (hunk 12 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenRequestDTO.java:11]: might be possible to leave it out, we just took it over. we would need to test that it does not break

File: src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java (hunk 10 above)
Comment (krusche) [reply src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java]: it was not working properly and there is already another check that all JsonInclude use the non empty version

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenListResponseDTO.java (hunk 11 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabPersonalAccessTokenListResponseDTO.java:14]: just reused from the original class based implementation. Can probably be removed, but we would need to test the change

File: src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java [hunk 14]
```diff
@@ -100,8 +102,8 @@ public void sendSubmissions(Exercise exercise, int maxRetries) {
      * @param submissions the submissions to send
      * @param maxRetries  number of retries before the request will be canceled
      */
-    public void sendSubmissions(Exercise exercise, Set<Submission> submissions, int maxRetries) {
-        Set<Submission> filteredSubmissions = new HashSet<>(submissions);
+    public void sendSubmissions(de.tum.in.www1.artemis.domain.Exercise exercise, Set<de.tum.in.www1.artemis.domain.Submission> submissions, int maxRetries) {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java:105]: In my opinion the change from `ExerciseDTOInterface` (and all similar occurrences) leads to code that is more difficult to read and understand: 

- here we have to write out the complete domain due to conflicting namespaces, which leads to very long names.
- In other classes we return objects (e.g. `Exercise`) that are actually a DTO and not an entity, but developers would only see that if they looked at the imports. 

Can you maybe elaborate what your motivation was behind changing this?

**Review by iyannsch — COMMENTED** [review]
Do you claim completeness with those changes, as in "there are no old/class DTOs anymore"? As far as I can tell, you transformed many class-DTOs to record-DTOs drastically reducing LOCs and increasing readability for many methods. However, as some other reviewers already pointed out, the need for lengthy classnames (and paths) is a clear downside of this approach in my opinion.

Keep in mind that I´m quite new to Artemis and I´m more than happy to get to know a simple or maybe even obvious answer to my idea as follows ;)
I´d like to raise the question of why we even need explicit DTOs all the time - let me give you an example:
The `ExamAttendanceCheckEvent`.`asDTO()` method exposes a DTO containing the `id`, `createdBy`, `createdDate`, and `textContent`. The `ExamAttendanceCheckEvent` class itself only contains the `textContent` itself and inherits `id`, `createdBy`, and `createdDate` from the `ExamLiveEvent` class. The new `ExamAttendanceCheckEventDTO` bundles those 4 properties.

But what if we could use the whole `ExamAttendanceCheckEvent` (including its parent-class) as a record and, thus, export it implicitly as a DTO? This would save a lot of transformation work, wouldn´t it? In Java, `record`s cannot extend nor be extended but they can implement interfaces. I think it would be possible to transform many data-classes (as `ExamAttendanceCheckEvent`) to records with the great advantage of reducing the number of classes and removing the difference between a DTO and its original object.

What do you think about that? Is there a design reason we opted for explicit DTOs in the whole system?

**Comment by rstief** [comment]
@iyannsch 
> the need for lengthy classnames (and paths) is a clear downside of this approach in my opinion.
The paths are not necessarily a problem of the migration to the DTOs and only occur if DTOs implement an interface with the same name as the class they are a DTO for. If we would rename those interfaces this problem no longer exist.

> I´d like to raise the question of why we even need explicit DTOs all the time - let me give you an example: The `ExamAttendanceCheckEvent`.`asDTO()` method exposes a DTO containing the `id`, `createdBy`, `createdDate`, and `textContent`. The `ExamAttendanceCheckEvent` class itself only contains the `textContent` itself and inherits `id`, `createdBy`, and `createdDate` from the `ExamLiveEvent` class. The new `ExamAttendanceCheckEventDTO` bundles those 4 properties.
This is a fair point, not all DTOs directly make sense. However, it would be cumbersome to always check if the returned entities contain unnecessary information and we want to avoid extreme cases such as [the example in the documentation].(https://docs.artemis.cit.tum.de/dev/guidelines/server.html#implications-of-not-using-dtos)

> But what if we could use the whole `ExamAttendanceCheckEvent` (including its parent-class) as a record and, thus, export it implicitly as a DTO? This would save a lot of transformation work, wouldn´t it? In Java, `record`s cannot extend nor be extended but they can implement interfaces. I think it would be possible to transform many data-classes (as `ExamAttendanceCheckEvent`) to records with the great advantage of reducing the number of classes and removing the difference between a DTO and its original object.
> 
> What do you think about that? Is there a design reason we opted for explicit DTOs in the whole system?
There are a few reasons why this approach would not be beneficial:
1. Records are immutable - This would make working with data objects very hard as we would always have to create a new record when we want to change a field.
2. I am not sure this change would even be possible (= compatible with Spring), especially for classes that currently inherit from others. 
3. Even if we could use records for our entities, we still have the problem that they often contain data we do not want to send over the network.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java (hunk 14 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java:105]: I agree that it's not ideal, but I would also avoid putting "DTO" and "Interface" into an interface. I am open to suggestions

File: src/main/java/de/tum/in/www1/artemis/service/TeamService.java [hunk 15]
```diff
@@ -69,20 +69,18 @@ public TeamService(TeamRepository teamRepository, UserRepository userRepository,
...
-        Map<Long, Long> userIdAndTeamIdMap = userIdAndTeamIdPairs.stream().collect(Collectors.toMap(userIdAndTeamIdPair -> userIdAndTeamIdPair[0], // userId
+        Map<Long, Long> userIdToTeamIdMap = userIdAndTeamIdPairs.stream().collect(Collectors.toMap(userIdAndTeamIdPair -> userIdAndTeamIdPair[0], // userId
                 userIdAndTeamIdPair -> userIdAndTeamIdPair[1] // teamId
         ));
 
         // Annotate to which team the user is already assigned to for the given exercise (null if not assigned)
-        teamSearchUsers.forEach(user -> user.setAssignedTeamId(userIdAndTeamIdMap.get(user.getId())));
-
-        return teamSearchUsers;
+        return users.stream().map(user -> new TeamSearchUserDTO(user.getId(), user.getLogin(), user.getName(), user.getFirstName(), user.getLastName(), user.getEmail(),
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/TeamService.java:82]: Yes, I agree. Would be cool if someone can help to change it

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/AssessmentUpdateDTO.java (hunk 6 above)
Comment (DominikRemo) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/AssessmentUpdateDTO.java]: Apparently, I misunderstood the expected behavior. The server correctly throws a BadRequestAlertException if the list is null and the test was wrong. I now fixed the test.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabCommitDTO.java (hunk 3 above)
Comment (DominikRemo) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabCommitDTO.java:20]: I couldn't find a place where we access these lists explicitly and the tests succeed without setting these default values. However, I am not familiar with the GitLab API and don't know if we can just remove the code. Maybe @krusche can provide more information

File: src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java (hunk 14 above)
Comment (JohannesWt) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSendingService.java:105]: Maybe it makes sense to call the Interface `ExerciseBaseDTO` and then the DTO `ExerciseDTO` so interfaces are always the Base for an DTO.

File: src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamLiveEvent.java [hunk 16]
```diff
@@ -81,11 +80,5 @@ public void setStudentExamId(Long studentExamId) {
...
     }
 
-    protected void populateDTO(ExamLiveEventDTO dto) {
-        dto.setId(getId());
-        dto.setCreatedBy(getCreatedBy());
-        dto.setCreatedDate(getCreatedDate());
-    }
-
-    public abstract ExamLiveEventDTO asDTO();
+    public abstract de.tum.in.www1.artemis.web.rest.dto.examevent.ExamLiveEvent asDTO();
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamLiveEvent.java:83]: Why did you change the name from `ExamLiveEventDTO` to `ExamLiveEvent`?
Imo the entity classes and DTOs should be quite easily distinguishable in code, since they are used for two different purposes.

Comment (DominikRemo) [reply src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamLiveEvent.java:83]: ExamLiveEventDTO has been renamed because it is an interface.

There is also [another discussion](https://github.com/ls1intum/Artemis/pull/8571#discussion_r1615315172) that suggests naming the interfaces for DTOs SomethingBaseDTO:
> How about calling the Interface `ExerciseBaseDTO` and then the DTO `ExerciseDTO` - so interfaces are always the Base for an DTO.

Feel free to join the discussion and to make suggestions :)

File: src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java (hunk 9 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java:32]: This currently uses inheritance and polymorphism, since asDTO() is called on objects that have the abstract super class as static type. This will be solved in a follow up

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 17]
```diff
@@ -92,8 +92,8 @@ public AssessmentService(ComplaintResponseService complaintResponseService, Comp
      * @param assessmentUpdate the assessment update containing a ComplaintResponse and the updated Feedback list
      * @return the updated Result
      */
-    public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exercise, AssessmentUpdate assessmentUpdate) {
-        if (assessmentUpdate.feedbacks() == null || assessmentUpdate.complaintResponse() == null) {
+    public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exercise, AssessmentUpdateBaseDTO assessmentUpdate) {
+        if (assessmentUpdate.complaintResponse() == null) {
             throw new BadRequestAlertException("Feedbacks and complaint response must not be null.", "AssessmentUpdate", "notnull");
         }
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:98]: Why was the check for feedbacks removed? If it was intentional please update the exception as follows:
```suggestion
    public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exercise, AssessmentUpdateBaseDTO assessmentUpdate) {
        if (assessmentUpdate.complaintResponse() == null) {
            throw new BadRequestAlertException("Complaint response must not be null.", "AssessmentUpdate", "notnull");
        }
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaDTOConverterService.java [hunk 18]
```diff
@@ -63,7 +60,7 @@ public AthenaDTOConverterService(TextBlockRepository textBlockRepository, TextEx
      * @param exercise the exercise to convert
      * @return *ExerciseDTO for Athena
      */
-    public ExerciseDTO ofExercise(Exercise exercise) {
+    public ExerciseBaseDTO ofExercise(de.tum.in.www1.artemis.domain.Exercise exercise) {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaDTOConverterService.java:63]: Since the DTOs were re-named, we should be able to import Exercise and the other entities to avoid using these long names.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSendingService.java [hunk 19]
```diff
@@ -59,7 +60,8 @@ private record ResponseDTO(String data) {
      * @param feedbacks  the feedback given by the tutor
      */
     @Async
-    public void sendFeedback(Exercise exercise, Submission submission, List<Feedback> feedbacks) {
+    public void sendFeedback(de.tum.in.www1.artemis.domain.Exercise exercise, de.tum.in.www1.artemis.domain.Submission submission,
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaFeedbackSendingService.java:63]: same here

File: src/main/java/de/tum/in/www1/artemis/service/dto/athena/Feedback.java [hunk 20]
```diff
@@ -6,5 +6,5 @@
  * Interface used to type the FeedbackDTOs for Athena: ProgrammingFeedbackDTO and TextFeedbackDTO
  */
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
-public interface FeedbackDTO {
+public interface Feedback {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/dto/athena/Feedback.java:9]: this is still Feedback -> should be FeedbackBaseDTO

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java [hunk 21]
```diff
@@ -572,8 +571,8 @@ public ResponseEntity<List<ExamLiveEventDTO>> getExamLiveEvents(@PathVariable Lo
 
         studentExamAccessService.checkCourseAndExamAccessElseThrow(courseId, examId, currentUser, studentExam.isTestRun(), false);
 
-        List<ExamLiveEventDTO> examLiveEvents = examLiveEventRepository.findAllByStudentExamIdOrGlobalByExamId(examId, studentExam.getId()).stream().map(ExamLiveEvent::asDTO)
-                .toList();
+        List<ExamLiveEventBaseDTO> examLiveEvents = examLiveEventRepository.findAllByStudentExamIdOrGlobalByExamId(examId, studentExam.getId()).stream()
+                .map(de.tum.in.www1.artemis.domain.exam.event.ExamLiveEvent::asDTO).toList();
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:575]: We can import ExamLiveEvent

**Review by laadvo — APPROVED** [review]
Tested participation and assessment of programming, text, modeling, and file upload exercises locally. Working as expected, did not encounter any null pointers or bad requests

**Review by rstief — APPROVED** [review]
Code LGTM, thanks for implementing my many change requests 🙏🙏🙏

**Review by DominikRemo — APPROVED** [review]
Tested on TS5. Empty commits (empty additions/deletions/completely empty commits) work for programming exercises and empty submissions for all other exercise types also work (tested as part of an exam)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabCommitDTO.java (hunk 3 above)
Comment (DominikRemo) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/dto/GitLabCommitDTO.java]: Tested on ts5. Works with empty commits after the code was removed

**Review by az108 — APPROVED** [review]
tested in a testing session on ts5. Participated in a exam and tried various exercises and different scenarios. All work as expected.

**Review by EneaGore — APPROVED** [review]
Tested on testing session ts5. Works as expected.

**Review by laadvo — APPROVED** [review]
Tested in the testing session on ts5, competencies are still working as expected

**Review by sarpsahinalp — APPROVED** [review]
Tested locally works as expected

**Review by JohannesWt — APPROVED** [review]
Reapprove - thanks for the changes
